File: ports/ts/src/chunker/chunker.ts
// The pinned CiteNexus recursive chunker (SPEC-PORTS-v1 §4). Parity with the
// Python reference citenexus.evidence.chunker.chunk_text.
//
// Recursive, boundary-aware splitting: keep whole paragraphs together; when a
// unit exceeds the token bound, recurse to the next finer boundary (paragraph ->
// line -> sentence -> word), then greedily pack pieces into size-bounded chunks
// with a trailing-word overlap window. "Token" count is whitespace-split words.
//
// RE2/lookbehind note: sentence splitting matches Python's (?<=[.!?])\s+ by
// scanning whitespace runs that immediately follow a . ! or ? and splitting
// there, keeping the punctuation on the left piece.

const PARAGRAPH = /\n\s*\n/;
const LINE = /\n/;
const WHITESPACE = /\s+/;

export const DEFAULT_MAX_TOKENS = 450;
export const DEFAULT_OVERLAP = 60;

function tokens(text: string): number {
  return text.split(WHITESPACE).filter((w) => w.length > 0).length;
}

/** Split on the coarsest boundary that yields more than one piece. */
function splitUnits(text: string): string[] {
  for (const pattern of [PARAGRAPH, LINE]) {
    const pieces = text
      .split(pattern)
      .map((p) => p.trim())
      .filter((p) => p.length > 0);
    if (pieces.length > 1) {
      return pieces;
    }
  }
  // Sentence boundary: whitespace runs immediately following . ! or ?.
  const sentences = splitSentences(text)
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
  if (sentences.length > 1) {
    return sentences;
  }
  // Finest boundary: individual words.
  return text.split(WHITESPACE).filter((w) => w.length > 0);
}

/** Mirror Python re.split(r"(?<=[.!?])\s+", text) without lookbehind. */
function splitSentences(text: string): string[] {
  const out: string[] = [];
  let start = 0;
  const re = /\s+/g;
  let m: RegExpExecArray | null;
  while ((m = re.exec(text)) !== null) {
    const prev = text[m.index - 1];
    if (prev === "." || prev === "!" || prev === "?") {
      out.push(text.slice(start, m.index));
      start = m.index + m[0].length;
    }
  }
  out.push(text.slice(start));
  return out;
}

/** Recursively split until every piece fits max_tokens. */
function fitPieces(text: string, maxTokens: number): string[] {
  if (tokens(text) <= maxTokens) {
    return [text];
  }
  const units = splitUnits(text);
  if (units.length === 1) {
    // A single oversized word-run: hard-split into max_tokens windows.
    const words = units[0]!.split(WHITESPACE).filter((w) => w.length > 0);
    const out: string[] = [];
    for (let i = 0; i < words.length; i += maxTokens) {
      out.push(words.slice(i, i + maxTokens).join(" "));
    }
    return out;
  }
  const out: string[] = [];
  for (const unit of units) {
    out.push(...fitPieces(unit, maxTokens));
  }
  return out;
}

/** The trailing pieces whose combined size is within overlap tokens. */
function overlapTail(pieces: string[], overlap: number): string[] {
  if (overlap <= 0) {
    return [];
  }
  const tail: string[] = [];
  let size = 0;
  for (let i = pieces.length - 1; i >= 0; i--) {
    const piece = pieces[i]!;
    const n = tokens(piece);
    if (size + n > overlap) {
      break;
    }
    tail.unshift(piece);
    size += n;
  }
  return tail;
}

/** Greedily pack pieces into chunks; carry an overlap tail between chunks. */
function pack(pieces: string[], maxTokens: number, overlap: number): string[] {
  const chunks: string[] = [];
  let current: string[] = [];
  let size = 0;
  for (const piece of pieces) {
    const n = tokens(piece);
    if (current.length > 0 && size + n > maxTokens) {
      chunks.push(current.join("\n"));
      const tail = overlapTail(current, overlap);
      current = tail.slice();
      size = current.reduce((acc, p) => acc + tokens(p), 0);
    }
    current.push(piece);
    size += n;
  }
  if (current.length > 0) {
    chunks.push(current.join("\n"));
  }
  return chunks;
}

/** Split `text` into bounded, overlapping, boundary-respecting chunks. */
export function chunkText(
  text: string,
  maxTokens: number = DEFAULT_MAX_TOKENS,
  overlap: number = DEFAULT_OVERLAP,
): string[] {
  if (maxTokens < 1) {
    throw new Error("max_tokens must be >= 1");
  }
  const stripped = text.trim();
  if (stripped.length === 0) {
    return [];
  }
  if (tokens(stripped) <= maxTokens) {
    return [stripped];
  }
  const clampedOverlap = Math.max(0, Math.min(overlap, maxTokens - 1));
  const pieces = fitPieces(stripped, maxTokens);
  return pack(pieces, maxTokens, clampedOverlap);
}

File: ports/go/chunker/chunker.go
// Package chunker is the pinned CiteNexus recursive chunker (SPEC-PORTS-v1 §4).
//
// Definition (frozen contract): a recursive, boundary-aware splitter. "Token"
// count is the whitespace-split word count. Defaults max_tokens=450,
// overlap=60. Parity with the Python reference
// citenexus.evidence.chunker.chunk_text.
//
// Boundary order (coarsest first): paragraph (/\n\s*\n/), line (/\n/), sentence
// (whitespace immediately following . ! or ?). A single oversized word-run is
// hard-split into consecutive max_tokens-word windows. Pieces are then greedily
// packed into max_tokens chunks joined by "\n", carrying a trailing whole-piece
// overlap window between chunks.
package chunker

import (
	"regexp"
	"strings"
	"unicode"
)

const (
	DefaultMaxTokens = 450
	DefaultOverlap   = 60
)

var paragraphRE = regexp.MustCompile(`\n\s*\n`)

// tokens counts whitespace-split words (Python str.split()).
func tokens(text string) int {
	return len(strings.Fields(text))
}

// splitSentences mirrors Python re.split(`(?<=[.!?])\s+`, text): it splits on
// each maximal whitespace run that immediately follows a . ! or ?, consuming the
// whitespace and keeping the punctuation on the left piece. RE2 has no
// lookbehind, so this is done by scanning.
func splitSentences(text string) []string {
	rs := []rune(text)
	var out []string
	start := 0
	i := 0
	for i < len(rs) {
		if unicode.IsSpace(rs[i]) {
			j := i
			for j < len(rs) && unicode.IsSpace(rs[j]) {
				j++
			}
			if i > 0 && (rs[i-1] == '.' || rs[i-1] == '!' || rs[i-1] == '?') {
				out = append(out, string(rs[start:i]))
				start = j
			}
			i = j
		} else {
			i++
		}
	}
	out = append(out, string(rs[start:]))
	return out
}

// stripDrop strips each piece and drops empties.
func stripDrop(pieces []string) []string {
	out := make([]string, 0, len(pieces))
	for _, p := range pieces {
		s := strings.TrimSpace(p)
		if s != "" {
			out = append(out, s)
		}
	}
	return out
}

// splitUnits splits text on the coarsest boundary that yields more than one
// non-empty piece; finest fallback is individual words.
func splitUnits(text string) []string {
	for _, pieces := range [][]string{
		stripDrop(paragraphRE.Split(text, -1)),
		stripDrop(strings.Split(text, "\n")),
		stripDrop(splitSentences(text)),
	} {
		if len(pieces) > 1 {
			return pieces
		}
	}
	return strings.Fields(text)
}

// fitPieces recursively splits text until every piece fits maxTokens.
func fitPieces(text string, maxTokens int) []string {
	if tokens(text) <= maxTokens {
		return []string{text}
	}
	units := splitUnits(text)
	if len(units) == 1 {
		words := strings.Fields(units[0])
		var out []string
		for i := 0; i < len(words); i += maxTokens {
			end := i + maxTokens
			if end > len(words) {
				end = len(words)
			}
			out = append(out, strings.Join(words[i:end], " "))
		}
		return out
	}
	var out []string
	for _, unit := range units {
		out = append(out, fitPieces(unit, maxTokens)...)
	}
	return out
}

// overlapTail returns the trailing pieces whose combined word-count is within
// overlap tokens.
func overlapTail(pieces []string, overlap int) []string {
	if overlap <= 0 {
		return nil
	}
	var tail []string
	size := 0
	for i := len(pieces) - 1; i >= 0; i-- {
		n := tokens(pieces[i])
		if size+n > overlap {
			break
		}
		tail = append([]string{pieces[i]}, tail...)
		size += n
	}
	return tail
}

// pack greedily packs pieces into maxTokens chunks joined by "\n", carrying an
// overlap tail between chunks.
func pack(pieces []string, maxTokens, overlap int) []string {
	chunks := []string{}
	var current []string
	size := 0
	for _, piece := range pieces {
		n := tokens(piece)
		if len(current) > 0 && size+n > maxTokens {
			chunks = append(chunks, strings.Join(current, "\n"))
			tail := overlapTail(current, overlap)
			current = append([]string(nil), tail...)
			size = 0
			for _, p := range current {
				size += tokens(p)
			}
		}
		current = append(current, piece)
		size += n
	}
	if len(current) > 0 {
		chunks = append(chunks, strings.Join(current, "\n"))
	}
	return chunks
}

// ChunkText splits text into bounded, overlapping, boundary-respecting chunks.
// It panics if maxTokens < 1 (the Python reference raises ValueError).
func ChunkText(text string, maxTokens, overlap int) []string {
	if maxTokens < 1 {
		panic("max_tokens must be >= 1")
	}
	text = strings.TrimSpace(text)
	if text == "" {
		return []string{}
	}
	if tokens(text) <= maxTokens {
		return []string{text}
	}
	if overlap < 0 {
		overlap = 0
	}
	if overlap > maxTokens-1 {
		overlap = maxTokens - 1
	}
	pieces := fitPieces(text, maxTokens)
	return pack(pieces, maxTokens, overlap)
}
